In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
data=pd.read_csv('mushrooms.csv')

In [3]:
data

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0        p         x           s         n       t    p               f   
1        e         x           s         y       t    a               f   
2        e         b           s         w       t    l               f   
3        p         x           y         w       t    p               f   
4        e         x           s         g       f    n               f   
...    ...       ...         ...       ...     ...  ...             ...   
8119     e         k           s         n       f    n               a   
8120     e         x           s         n       f    n               a   
8121     e         f           s         n       f    n               a   
8122     p         k           y         n       f    y               f   
8123     e         x           s         n       f    n               a   

     gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0               c         n          k  ...                        s   
1               c         b          k  ...                        s   
2               c         b          n  ...                        s   
3               c         n          n  ...                        s   
4               w         b          k  ...                        s   
...           ...       ...        ...  ...                      ...   
8119            c         b          y  ...                        s   
8120            c         b          y  ...                        s   
8121            c         b          n  ...                        s   
8122            c         n          b  ...                        k   
8123            c         b          y  ...                        s   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                         w                      w         p          w   
1                         w                      w         p          w   
2                         w                      w         p          w   
3                         w                      w         p          w   
4                         w                      w         p          w   
...                     ...                    ...       ...        ...   
8119                      o                      o         p          o   
8120                      o                      o         p          n   
8121                      o                      o         p          o   
8122                      w                      w         p          w   
8123                      o                      o         p          o   

     ring-number ring-type spore-print-color population habitat  
0              o         p                 k          s       u  
1              o         p                 n          n       g  
2              o         p                 n          n       m  
3              o         p                 k          s       u  
4              o         e                 n          a       g  
...          ...       ...               ...        ...     ...  
8119           o         p                 b          c       l  
8120           o         p                 b          v       l  
8121           o         p                 b          c       l  
8122           o         e                 w          v       l  
8123           o         p                 o          c       l  

[8124 rows x 23 columns]

In [4]:
data.isna().sum()

class                       0
cap-shape                   0
cap-surface                 0
cap-color                   0
bruises                     0
odor                        0
gill-attachment             0
gill-spacing                0
gill-size                   0
gill-color                  0
stalk-shape                 0
stalk-root                  0
stalk-surface-above-ring    0
stalk-surface-below-ring    0
stalk-color-above-ring      0
stalk-color-below-ring      0
veil-type                   0
veil-color                  0
ring-number                 0
ring-type                   0
spore-print-color           0
population                  0
habitat                     0
dtype: int64

In [5]:
data.shape

(8124, 23)

In [6]:
data.nunique()

class                        2
cap-shape                    6
cap-surface                  4
cap-color                   10
bruises                      2
odor                         9
gill-attachment              2
gill-spacing                 2
gill-size                    2
gill-color                  12
stalk-shape                  2
stalk-root                   5
stalk-surface-above-ring     4
stalk-surface-below-ring     4
stalk-color-above-ring       9
stalk-color-below-ring       9
veil-type                    1
veil-color                   4
ring-number                  3
ring-type                    5
spore-print-color            9
population                   6
habitat                      7
dtype: int64

In [7]:
data.describe()

class cap-shape cap-surface cap-color bruises  odor gill-attachment  \
count   8124      8124        8124      8124    8124  8124            8124   
unique     2         6           4        10       2     9               2   
top        e         x           y         n       f     n               f   
freq    4208      3656        3244      2284    4748  3528            7914   

       gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
count          8124      8124       8124  ...                     8124   
unique            2         2         12  ...                        4   
top               c         b          b  ...                        s   
freq           6812      5612       1728  ...                     4936   

       stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
count                    8124                   8124      8124       8124   
unique                      9                      9         1          4   
top                         w                      w         p          w   
freq                     4464                   4384      8124       7924   

       ring-number ring-type spore-print-color population habitat  
count         8124      8124              8124       8124    8124  
unique           3         5                 9          6       7  
top              o         p                 w          v       d  
freq          7488      3968              2388       4040    3148  

[4 rows x 23 columns]

In [8]:
data['class'].value_counts()

e    4208
p    3916
Name: class, dtype: int64

In [9]:
data['veil-type'].value_counts()

p    8124
Name: veil-type, dtype: int64

Since, all the records (8124) for the feature 'veil-type' is same, it does not contribute anything to the dataset. Therefore, dropping the 'veil-type' feature. 

In [10]:
data.drop(['veil-type'], axis=1, inplace=True)

In [11]:
data.head()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-above-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-surface-below-ring stalk-color-above-ring stalk-color-below-ring  \
0                        s                      w                      w   
1                        s                      w                      w   
2                        s                      w                      w   
3                        s                      w                      w   
4                        s                      w                      w   

  veil-color ring-number ring-type spore-print-color population habitat  
0          w           o         p                 k          s       u  
1          w           o         p                 n          n       g  
2          w           o         p                 n          n       m  
3          w           o         p                 k          s       u  
4          w           o         e                 n          a       g  

[5 rows x 22 columns]

#### Encoding the categorical features to numeric values

In [12]:
from sklearn.preprocessing import LabelEncoder
encode = LabelEncoder()
col = data.columns
for i in col:
    data[i]=encode.fit_transform(data[i])

In [13]:
data

class  cap-shape  cap-surface  cap-color  bruises  odor  \
0         1          5            2          4        1     6   
1         0          5            2          9        1     0   
2         0          0            2          8        1     3   
3         1          5            3          8        1     6   
4         0          5            2          3        0     5   
...     ...        ...          ...        ...      ...   ...   
8119      0          3            2          4        0     5   
8120      0          5            2          4        0     5   
8121      0          2            2          4        0     5   
8122      1          3            3          4        0     8   
8123      0          5            2          4        0     5   

      gill-attachment  gill-spacing  gill-size  gill-color  ...  \
0                   1             0          1           4  ...   
1                   1             0          0           4  ...   
2                   1             0          0           5  ...   
3                   1             0          1           5  ...   
4                   1             1          0           4  ...   
...               ...           ...        ...         ...  ...   
8119                0             0          0          11  ...   
8120                0             0          0          11  ...   
8121                0             0          0           5  ...   
8122                1             0          1           0  ...   
8123                0             0          0          11  ...   

      stalk-surface-above-ring  stalk-surface-below-ring  \
0                            2                         2   
1                            2                         2   
2                            2                         2   
3                            2                         2   
4                            2                         2   
...                        ...                       ...   
8119                         2                         2   
8120                         2                         2   
8121                         2                         2   
8122                         2                         1   
8123                         2                         2   

      stalk-color-above-ring  stalk-color-below-ring  veil-color  ring-number  \
0                          7                       7           2            1   
1                          7                       7           2            1   
2                          7                       7           2            1   
3                          7                       7           2            1   
4                          7                       7           2            1   
...                      ...                     ...         ...          ...   
8119                       5                       5           1            1   
8120                       5                       5           0            1   
8121                       5                       5           1            1   
8122                       7                       7           2            1   
8123                       5                       5           1            1   

      ring-type  spore-print-color  population  habitat  
0             4                  2           3        5  
1             4                  3           2        1  
2             4                  3           2        3  
3             4                  2           3        5  
4             0                  3           0        1  
...         ...                ...         ...      ...  
8119          4                  0           1        2  
8120          4                  0           4        2  
8121          4                  0           1        2  
8122          0                  7           4        2  
8123          4                  4           1        2  

[8124 rows x 22 columns]

#### Seperating target and independent features

In [14]:
X = data.drop(['class'], axis=1)
y = data['class']

In [15]:
X

cap-shape  cap-surface  cap-color  bruises  odor  gill-attachment  \
0             5            2          4        1     6                1   
1             5            2          9        1     0                1   
2             0            2          8        1     3                1   
3             5            3          8        1     6                1   
4             5            2          3        0     5                1   
...         ...          ...        ...      ...   ...              ...   
8119          3            2          4        0     5                0   
8120          5            2          4        0     5                0   
8121          2            2          4        0     5                0   
8122          3            3          4        0     8                1   
8123          5            2          4        0     5                0   

      gill-spacing  gill-size  gill-color  stalk-shape  ...  \
0                0          1           4            0  ...   
1                0          0           4            0  ...   
2                0          0           5            0  ...   
3                0          1           5            0  ...   
4                1          0           4            1  ...   
...            ...        ...         ...          ...  ...   
8119             0          0          11            0  ...   
8120             0          0          11            0  ...   
8121             0          0           5            0  ...   
8122             0          1           0            1  ...   
8123             0          0          11            0  ...   

      stalk-surface-above-ring  stalk-surface-below-ring  \
0                            2                         2   
1                            2                         2   
2                            2                         2   
3                            2                         2   
4                            2                         2   
...                        ...                       ...   
8119                         2                         2   
8120                         2                         2   
8121                         2                         2   
8122                         2                         1   
8123                         2                         2   

      stalk-color-above-ring  stalk-color-below-ring  veil-color  ring-number  \
0                          7                       7           2            1   
1                          7                       7           2            1   
2                          7                       7           2            1   
3                          7                       7           2            1   
4                          7                       7           2            1   
...                      ...                     ...         ...          ...   
8119                       5                       5           1            1   
8120                       5                       5           0            1   
8121                       5                       5           1            1   
8122                       7                       7           2            1   
8123                       5                       5           1            1   

      ring-type  spore-print-color  population  habitat  
0             4                  2           3        5  
1             4                  3           2        1  
2             4                  3           2        3  
3             4                  2           3        5  
4             0                  3           0        1  
...         ...                ...         ...      ...  
8119          4                  0           1        2  
8120          4                  0           4        2  
8121          4                  0           1        2  
8122          0                  7           4        2  
8123          4 

In [16]:
y

0       1
1       0
2       0
3       1
4       0
       ..
8119    0
8120    0
8121    0
8122    1
8123    0
Name: class, Length: 8124, dtype: int32

#### Splitting into test and train set

In [17]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(X,y, test_size=0.20, random_state=42)

### Decision Tree

In [18]:
from sklearn.tree import DecisionTreeClassifier
dt_class = DecisionTreeClassifier()
dt_class.fit(x_train, y_train)

DecisionTreeClassifier()

In [19]:
y_pred_dt = dt_class.predict(x_test)

In [20]:
from sklearn.metrics import confusion_matrix, accuracy_score
confusion_matrix(y_test, y_pred_dt)

array([[843,   0],
       [  0, 782]], dtype=int64)

In [21]:
accuracy_score(y_test, y_pred_dt)

1.0

### Logistic Regression

In [22]:
from sklearn.linear_model import LogisticRegression
log_R = LogisticRegression(max_iter=500, )
log_R.fit(x_train, y_train)
y_pred_lr =log_R.predict(x_test) 

In [23]:
confusion_matrix(y_pred_lr,y_test)

array([[799,  41],
       [ 44, 741]], dtype=int64)

In [24]:
accuracy_score(y_pred_lr,y_test)

0.9476923076923077

In [25]:
from sklearn.metrics import classification_report
print(classification_report(y_pred_lr, y_test))

              precision    recall  f1-score   support

           0       0.95      0.95      0.95       840
           1       0.95      0.94      0.95       785

    accuracy                           0.95      1625
   macro avg       0.95      0.95      0.95      1625
weighted avg       0.95      0.95      0.95      1625



### SVM Classifier

In [26]:
from sklearn.svm import SVC
svm = SVC(gamma='scale')
svm.fit(x_train, y_train)
y_pred_sc = svm.predict(x_test)

In [27]:
accuracy_score(y_pred_sc, y_test)

0.9938461538461538

In [28]:
confusion_matrix(y_pred_sc, y_test)

array([[842,   9],
       [  1, 773]], dtype=int64)

### KNN Classifier

In [29]:
from sklearn.neighbors import KNeighborsClassifier
for i in range(1,10):
    knn = KNeighborsClassifier(n_neighbors=i)
    knn.fit(x_train, y_train)
    y_t=knn.predict(x_test)
    print((accuracy_score(y_t,y_test)*100))

99.93846153846154
99.93846153846154
99.81538461538462
99.63076923076923
99.63076923076923
99.6923076923077
99.56923076923077
99.50769230769231
99.50769230769231


Value of K=1 or K=2 gives the best accuracy

In [30]:
knn = KNeighborsClassifier(n_neighbors=2)
knn.fit(x_train, y_train)
y_pred_knn=knn.predict(x_test)
print((accuracy_score(y_pred_knn,y_test)*100))

99.93846153846154


In [31]:
confusion_matrix(y_pred_knn,y_test)

array([[842,   0],
       [  1, 782]], dtype=int64)

### Naive Bayes Classifier

In [32]:
from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()
gnb.fit(x_train, y_train)
y_pred_gnb = gnb.predict(x_test)

In [33]:
accuracy_score(y_pred_gnb,y_test)

0.9218461538461539

In [34]:
confusion_matrix(y_pred_gnb, y_test)

array([[771,  55],
       [ 72, 727]], dtype=int64)

In [35]:
from sklearn.naive_bayes import BernoulliNB
bnb = BernoulliNB()
bnb.fit(x_train, y_train)
y_pred_bnb = bnb.predict(x_test)

In [36]:
accuracy_score(y_pred_bnb,y_test)

0.8418461538461538

In [37]:
confusion_matrix(y_pred_bnb,y_test)

array([[797, 211],
       [ 46, 571]], dtype=int64)